### PR TITLE
Basic index functionality

### PR DESCRIPTION
**Pull request by carlthuringer**
@bellycard/platform This is the initial functional prototype of Gladius, the tool for cutting through JSONAPI.

**Why does this exist?** Because [All existing clients](http://jsonapi.org/implementations/#client-libraries-ruby) follow an ORM-like pattern of defining a Class, which inherits from some base class, which implements the functionality. I dont' want to have to define classes and initializers, I just want to talk to a REST API, is that so hard?!

Things it doesn't yet support:
* Caching Resource structures
* Pagination
* Links/hypermedia
* Relationships

Things it does support:
* Index
* Get
* Create
* Update

I guess I didn't implement delete yet.

I'm looking for feedback, and a deputy to help maintain and move this forward. I hope this could be the basis of internal clients consuming our APIs, as well as being useful for our own tooling.

Cool Stuff:
* The Smallest Rails App Ever (just for testing, in-memory sqlite!)
* Faraday Rack adapter, which lets you make requests to rack applications, exercising more of the stack. Nice alternative *(compliment?)* to `Rack::Test`.
* Pretty printer.

Yes, I know:
* Not much test coverage, or consistency.
* Method missing (I tried not to, but this is better than subclassing Struct)

### CODE REVIEW

File: spec/support/test_app.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+end
+
+# posts resource
+class PostResource < JSONAPI::Resource
+  attribute :title
+  attribute :body
+end
+
+# the core application
+class TestApp < Rails::Application
```
Comment (jonhoman) [inline spec/support/test_app.rb:42]: This is an interesting way to write a Rails app.

**Review by jonhoman — COMMENTED** [review]
Before we continue to push forward on this effort, I think the team should consider whether this is a project we think is worthwhile at our current size. Basically the question is: should a team of our size write our own JSONAPI client instead of using existing tools?

**Comment by carlthuringer** [comment]
My argument for this is that developer time spent writing lines like
```ruby
before do
request.headers["Content-Type"] = "application/json.api+vnd"
end
```
and
```ruby
data = JSON.parse(response.body, symbolize_name: true)[:data]
```
can be spent writing tools that eliminate such bugs and offer greater efficiency when interacting with JSON apis.

We could distribute or copypasta a handful of helpers to do the above blocks of code, and fix the errors I typed in there, or we could distribute a tool that works just as well for testing as it does for controllers, models, service objects and IRB.

**Comment by carlthuringer** [comment]
@sisk @jonhoman Any further opinions on this? If not, I'm going to just close this up and abandon.

**Comment by sisk** [comment]
I very honestly just have not had the free time to digest it. Can we just leave it open for a while?

**Review by sisk — APPROVED** [review]
I think there is some merit to having a tool like this, which helps bridge the gap between cumbersome JSONAPI interactions and an overly-prescriptive thing like service-client.

Apologies for taking so long to review it.

**Pull request merged**